devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
     def __init__(self, *args, **kwargs):
         if not self._cached():
             self.name = kwargs.get('name')
-            self.shape = kwargs.get('shape', None)
-            if self.shape is None:
-                dimensions = kwargs.get('dimensions')
-                self.shape = tuple([d.size for d in dimensions])
+            self.grid = kwargs.get('grid', None)
+
+            if self.grid is None:
+                self.shape_domain = kwargs.get('shape', None)
+                self.dtype = kwargs.get('dtype', np.float32)
+                if self.shape_domain is None:
+                    error("Creating a Function requires either 'shape'"
+                          "or a 'grid' argument")
+                    raise ValueError("Unknown symbol dimensions or shape")
```

Comment:
`self.grid = Grid(kwargs.get('shape', None))` with raise exception?

May lead to issues if using different dimensions for different objects.

Reply:
If you're using different dimensions for two objects you should either create tow-different grids and make sure they agree on all dimensions you want them to agree on, or explicitly provide `dimensions` and `shape` when creating your function objects.

Also, not providing a `shape` to a `Grid`, which ultimately defines the base of the spatial data discretisation within the physical domain, simply doesn't make sense to me.